token_examples/factory_token/src/token.rs

```diff
@@ -0,0 +1,419 @@
+            return Err(RuntimeError::AddressNotAuthorized);
+        }
+
+        let mut hook = self.token().mint(
+            &caller,
+            &params.initial_owner,
+            &params.amount,
+            params.operator_data,
+            Default::default(),
+        )?;
+
+        let cid = self.save()?;
+        sdk::sself::set_root(&cid).unwrap();
+
+        let hook_ret = hook.call(self.token().msg())?;
+
+        self.reload(&cid)?;
+        let ret = self.token().mint_return(hook_ret)?;
+
+        Ok(ret)
+    }
+
+    /// Permanently disable minting
+    /// Only the authorised mint operator can do this
+    ///
+    /// TODO: should authorised minter be separate to a token 'owner' ?
+    /// also, should it be a single address, a list of addresses, or some other means?
+    pub fn disable_mint(&mut self) -> Result<(), RuntimeError> {
+        let caller = caller_address();
+        // no minter means minting has already been permanently disabled
+        // we return this if already disabled because it will make more sense than failing the address check below
+        let minter = self.minter.ok_or(RuntimeError::MintingDisabled)?;
+        let msg = FvmMessenger::default();
+        if !msg.same_address(&caller, &minter) {
+            return Err(RuntimeError::AddressNotAuthorized);
+        }
+
+        self.minter = None;
+        Ok(())
+    }
+}
+
+pub fn deserialize_params<O: DeserializeOwned>(params: u32) -> O {
+    let params = sdk::message::params_raw(params).unwrap().1;
+    let params = RawBytes::new(params);
+    params.deserialize().unwrap()
+}
+
+pub fn return_ipld<T>(value: &T) -> std::result::Result<u32, RuntimeError>
+where
+    T: Serialize + ?Sized,
+{
+    let bytes = fvm_ipld_encoding::to_vec(value)?;
+    Ok(sdk::ipld::put_block(DAG_CBOR, bytes.as_slice())?)
+}
+
+/// Generic invoke for FRC46 Token methods
+/// Given a method number and parameter block id, invokes the appropriate method on the FRC46Token interface
+///
+/// The flush_state function passed into this must flush current state to the blockstore and update the root cid
+/// This is called after operations which mutate the state, such as changing an allowance or burning tokens.
+///
+/// Transfer and TransferFrom operations invoke the receiver hook which will require flushing state before calling the hook
+/// This must be done inside the FRC46Token::transfer/transfer_from functions
+///
+/// Possible returns:
+/// - Ok(None) - method not found
+/// - Ok(Some(u32)) - block id of results saved to blockstore (or NO_DATA_BLOCK_ID if there is no result to return)
+/// - Err(error) - any error encountered during operation
+///
+/// TODO: improve the error type/handling so this can be moved into a library and work with any token code 
+///
+pub fn frc46_invoke<T, F>(
+    method_num: u64,
+    params: u32,
+    token: &mut T,
+    flush_state: F,
+) -> Result<Option<u32>, RuntimeError>
+where
+    T: FRC46Token<RuntimeError>,
```

Comment:
Done. I made the return data serialization abort on errors to avoid the minefield of it being compatible with whatever error type a user implements. It should be in a state where it could be used in external code now.